Remora.Commands/Trees/CommandTreeBuilder.cs

```diff
@@ -190,14 +277,114 @@ private IEnumerable<CommandNode> GetModuleCommands(Type moduleType, IParentNode
-                moduleType,
-                method,
-                commandAttribute.Aliases
+                CreateDelegate(method, method.GetParameters().Select(p => p.ParameterType).ToArray()),
+                CommandShape.FromMethod(method),
+                commandAttribute.Aliases,
+                attributes,
+                conditions
             );
         }
     }
+
+    private static Func<IServiceProvider, object?[], CancellationToken, ValueTask<IResult>> CreateDelegate(MethodInfo method, Type[] argumentTypes)
+    {
+        // Get the object from the container
+        var serviceProvider = Expression.Parameter(typeof(IServiceProvider), "serviceProvider");
+
+        var instance = Expression.Call(serviceProvider, GetServiceMethodInfo, Expression.Constant(method.DeclaringType));
+
+        var parameters = Expression.Parameter(typeof(object?[]), "parameters");
+
+        // Create the arguments
+        var arguments = new Expression[argumentTypes.Length];
+        for (var i = 0; i < argumentTypes.Length; i++)
+        {
+            var argumentType = argumentTypes[i];
+            var argument = Expression.ArrayIndex(parameters, Expression.Constant(i));
+            arguments[i] = Expression.Convert(argument, argumentType);
+        }
+
+        var castedInstance = Expression.Convert(instance, method.DeclaringType!);
+
+        var call = Expression.Call(castedInstance, method, arguments);
+
+        // Convert the result to a ValueTask<IResult>
+        call = (MethodCallExpression)CoerceToValueTask(call);
+
+        var ct = Expression.Parameter(typeof(CancellationToken), "cancellationToken");
+
+        var block = Expression.Block
+        (
+            Expression.Call(castedInstance, SetCancellationTokenMethodInfo, ct),
+            call
+        );
+
+        // Compile the expression
+        var lambda = Expression.Lambda<Func<IServiceProvider, object?[], CancellationToken, ValueTask<IResult>>>(block, serviceProvider, parameters, ct);
+
+        return lambda.Compile();
+    }
+
+    /// <summary>
+    /// Coerces the static result type of an expression to a <see cref="ValueTask"/>.
+    /// <list type="bullet">
+    /// <item>If the type is <see cref="ValueTask"/>, returns the expression as-is</item>
+    /// <item>If the type is <see cref="Task"/>, returns an expression wrappiung the Task in a <see cref="ValueTask"/></item>
+    /// <item>If the type is <c>void</c>, returns <see cref="ValueTask.CompletedTask"/></item>
+    /// <item>Otherwise, throws <see cref="InvalidOperationException"/></item>
+    /// </list>
+    /// </summary>
+    /// <param name="expression">The input expression.</param>
+    /// <param name="expressionType">The type of the expression; defaults to <see cref="Expression.Type"/>.</param>
+    /// <returns>The new expression.</returns>
+    /// <exception cref="InvalidOperationException">If the type of <paramref name="expression"/> is not wrappable.</exception>
```

Comment:
Documentation is wrong (yes I know I wrote it, please fix it)

Reply:
I take it you're referring to `to a <see cref="ValueTask"/>.` where it should be `to a <see cref="ValueTask{IResult}"/>.`, but you're commenting on 11 lines of code so I don't know what's entirely wrong outside of that

Reply:
`void` return type does not return `ValueTask.CompletedTask`, it throws instead, that is a carryover from my old code

Reply:
`If the type is <see cref="ValueTask"/>` -> ValueTask<IResult>

etc.